torch/_functorch/_aot_autograd/collect_metadata_analysis.py

```diff
@@ -91,16 +99,62 @@ def coerce_tangent(x):
         out, "__coerce_tangent_metadata__"
     ):
         out = out.__coerce_tangent_metadata__()  # type: ignore[attr-defined]
-    # It's possible to have a subclass that advertises as contiguous,
-    # but has noncontiguous inner tensors.
-    # Force these to be conntiguous too
+
+    # Coerce subclasses elements to memory_format too.
     if is_traceable_wrapper_subclass(out):
-        for attr in out.__tensor_flatten__()[0]:  # type: ignore[attr-defined]
+        attrs = out.__tensor_flatten__()[0]
+
+        assert isinstance(memory_format, list)
+        for i, attr in enumerate(attrs):
+            elem = getattr(out, attr)
+            new_elem, updated = coerce_tangent(elem, memory_format[1 + i])
+            if updated:
+                setattr(out, attr, new_elem)
+
+    return out, updated
+
+
+# Coercing and collecting traced tangents memory format in one recursive traversal
+# mypy: ignore-errors
+def coerce_tangent_and_suggest_memory_format(
+    x: Tensor, force_memory_format: Optional[torch.memory_format] = None
+):
+    updated = False
+    if not isinstance(x, Tensor):
+        return x, None, updated
```

Comment:
this fn feels a bit complicated (it returns `[Tensor, MemoryFormat, bool]`?) - can you add return types to the fn and document them?

Reply:
reading this more - I think returning the `was_updated` bool is kind of confusing, since it is ignored in most places. What if instead, we:

- don't return a `was_updated` bool
- in the places where the caller needed it, they can just do the check themselves to know if contiguous had an effect:
```
updated_out, mem_format = coerce_tangent_and_suggest_memory_format(out)
if updated_out is not out:
    setattr(...)
```